Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -264,13 +281,16 @@ extension SwiftLanguageServer {
       // Already logged failure.
       return
     }
-    self.publishDiagnostics(response: dict, for: snapshot)
+    self.publishDiagnostics(
+        response: dict, for: snapshot, compileCommand: compileCmd)
   }
 
-  public func documentUpdatedBuildSettings(_ uri: DocumentURI, settings: FileBuildSettings?) {
+  public func documentUpdatedBuildSettings(_ uri: DocumentURI, change: FileBuildSettingsChange) {
     self.queue.async {
-      let compileCommand = settings.map { SwiftCompileCommand($0) }
+      let compileCommand = SwiftCompileCommand(change: change)
       // Confirm that the compile commands actually changed, otherwise we don't need to do anything.
+      // This includes when the compiler arguments are the same but the command is no longer
+      // considered to be fallback.
```

Comment:
If we really do have the same settings, why rebuild the AST instead of just updating the filtering?